### DISCUSSION TITLE
Big delay before first instruction for a swarm using Crazyradio PA

### DISCUSSION THREAD

**Discussion by MeghanHailey** [bitcraze/discussions#798] [Q&A] [answered]
Hello,
I am quite new to Crazyflies and have been working my way up through the step-by-step tutorials. I have individuals drones working just fine and can get up to 4 work in a swarm using the [tutorial code](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/step-by-step/sbs_swarm.py). However when I am working with a swarm there seems to be a large delay before the first instruction inside the `with Swarm` statement, in this case the light check. This delay seems to be about 10-15 seconds per drone that I am connected to and the instruction only gets executed for all of them after this cummulative delay. After I go above 4 drones this takes so long the instruction just times out and I get an error. When I remove the light check I see the same behaviour for the `reset_estimators()` command. I am using one Crazyradio PA to control a swarm of Crazyflie 2.1s. I have seen discussions about limitations of using just one radio to control them if you need to be sending a lot of data, but the script mentioned only sends a couple of instructions so I am not sure if this could be an issue. I have however realised some of those might have been about the newer Crazyraadio 2.0 and I was wondering if these issues could be caused by me using the older Crazyradio PA. I would be able to purchase the 2.0 version if you think this could improve my situation, but I haven't been able to find much information to help me judge whether this would improve my situation. Before purchasing the Crazyradio 2.0 I wanted to check if anyone thinks there is something else this could be caused by or any suggestions for what to try.
Many thanks,
Meghan

**Comment by krichardsson** [comment]
Hi,
there are multiple aspects of using many Crazyflies with one radio and I'll try to clear some of them.

The first thing to understand it that when the python lib connects to the Crazyflie there are a number of steps performed in the connection sequence, and some of them include downloading lists of the log and parameters in the Crazyflie (called TOC). When downloaded the TOC is stored in a file in a directory called `cache`, and the next time you connect the file will be used instead of downloading the TOC, which speeds the connection process up. When I tested the tutorial script on my machine, the first connection took around 30 seconds, while the second (with cached TOCs) took around 8. 
In the tutorial script the `CachedCfFactory` class is used to set up the caching and if you have not changed anything it should work. It is possible that the script does not have the right to write files (depending on your setup) and this would break the cache functionality. I think this might what you see.

Secondly when the text `Connected to  Crazyflies` is displayed, this is only partially true. There is enough information available to do most stuff, but there is still some more parameter data to be downloaded and this is blocking the first call in `light_check()` a bit longer, as this call sets a parameter. In my setup I got the connected message after 3 seconds when the TOCs are cached.

Thirdly the Crazyradio and the communication system was really designed for only one Crazyflie, and when extending it to use multiple CFs there is unfortunately overhead, that is connecting to 4 Crazyflies will take longer than 4 connections to one Crazyflie. 

The radio channels/addresses used will also effect the communication efficiency. It is better to use the same radio channel for all Crazyflies with different addresses (as in the tutorial) than the other way around. 

Finally there might be interference on the radio channel you use which will trigger re-sends of packets. Look at the LED on the Crazyradio, it will be red when packets are lost. If you get a lot of red it might be worth trying to improve reception or try another channel.

Currently there is no difference between the Crazyradio PA and Crazyradio 2.0 as they use the same type of communication. The Crazyradio 2.0 will enable us to rewrite the communication stack though, and this should solve most of the problems we see today. This is something for the future and right now a Crazyradio 2.0 will unfortunately not make any difference.

**Reply by MeghanHailey** [reply]
Hi, thanks so much for the detailed reply and explaining everything.

I am running my setup on the provided Bitcraze virtual machine, but I am not very experienced with Linux persmissions. Is there any way I could check if the permissions are the issue and if the cache is being created? Where would the `cache` directory be located? There seems to be no directory where the Python code file is stored. I could find such directory in the home/bitcraze folder, however wasn't able to identify if any of the files there are relevant.

I am using the same radio channel for all of them, but I will look into trying different channels. The LED for the radio is green but also pretty much constantly flashing red, but didn't realise this indicated packet loss as the getting started tutorial says "Radio connected: The front left LED (4) is flickering in red and/or green." which made me think the red has no significance, but I understand this is just a basic tutorial and not all detail was included.

I also hope to try using multiple radios later today to see if I can improve this, but would be great if we can work out a way to reduce the time it takes with with one radio as I would still be hoping to control at least 3 or 4 with one radio.

And thanks for clarifying the Crazyradio 2.0 wouldn't help currently.

**Reply by MeghanHailey** [reply]
I did manage to get 6 of them to fly together today using 2 radios and since I am trying to fly 10 of them and have 3 radios this should work, however the delay is still annoyingly long, so would like to look into whether the cache is working correctly and would appreciate any guidance on that as I seem to be getting longer times than you do for the second connection (and I think it might have been the same time for me for first connection).

I have also tried switching between many different channels through the whole range with little to no difference and there seems to still be a lot of red on the radio LED. Is there anything that can be done about this or is this just interference from my surroundings?

**Comment by krichardsson** [comment]
Hi,
the cache folder should be created in the working directory. If you start your script from a terminal, it should be the directory that you are "located in" in the terminal. If you run the script from VScode, it is usually the the root of the file tree.
If you are using the VM and are running somewhere in the bitcraze file tree (the default) it should work.

If you use more than one radio, make sure they use different channels. The Crazyflies on one radio is using one channel (but different addresses), the crazyflies on the next radio is on a different channel (with some set of addresses) and so on.

For instance:

Radio 1:
radio://0/30/2M/E7E7E7E7E7
radio://0/30/2M/E7E7E7E7E8
radio://0/30/2M/E7E7E7E7E9 

Radio 2:
radio://0/50/2M/E7E7E7E7E7
radio://0/50/2M/E7E7E7E7E8
radio://0/50/2M/E7E7E7E7E9 

Radio 3:
radio://0/70/2M/E7E7E7E7E7
radio://0/70/2M/E7E7E7E7E8
radio://0/70/2M/E7E7E7E7E9

**Reply by MeghanHailey** [reply]
Thanks for the explanation. I was able to locate the cache folder in the root of the file tree when running from VSCode and it was also generated where the script was run when trying to run it from the terminal, so it seems the cache files are being created. 

I have been using the radios on a different channel each as suggested, do you need to use the same set of addresses for each radio as you described or does this not matter?

Thanks.

**Reply by MeghanHailey** [reply]
I tried to do a test today and for 6 drones on 2 radios (each using a different channel) I am getting around 70 seconds from starting the code to them finishing the light check and resetting estimators. This seems to be about the same each run and regardless of if I run it in VSCode or in terminal.

**Comment by krichardsson** [comment]
> I have been using the radios on a different channel each as suggested, do you need to use the same set of addresses for each radio as you described or does this not matter?

Just to make sure we use the same words: The channel is setting the frequency that the radio is using, this is the `30`, `50` and `70` in the example above. The address is encoded in the radio packets, this is the `E7E7E7E7E7`, `E7E7E7E7E8` and so on.

The addresses do not have to be the same between radios (I just copy/pasted them), this is an id that is part of the data in the radio packet that is used by the Crazyflie to determine if a certain packet should be used or not. The channel (the radio frequency), on the other hand, takes a while to change in the radio and it is more efficient to use the same channel for all transmissions on one Crazyradio. 
If the same channel (frequency) is used by two different Crazyradios, there will be collisions when both of them transmit at the same time, and this should be avoided to minimize packet loss.

**Reply by MeghanHailey** [reply]
Thank you, it seems I understood correctly as I am using a different channel for each of the radios. Since the cache seems to be saved correctly (I am attaching [my cache files](https://github.com/bitcraze/discussions/files/11959224/cache.zip) to check this is expected files that should be saved) and I am still getting over a minute for 3 drones per radio. Do you think this is an issue with the packet loss since the radio lights are flashing green and red all the time? Is there anything I can do about this or is this just a factor of my environement? I also tried at home to see if where I normally test has a lot of noise but the light flashing seemed about the same to me. I also occasionally get disconnected when connected to a single drone via cfclient, so maybe there is an issue with the signal here. Or do you have any suggestions in general about what I could try to reduce the startup delay?
Thank you.

**Reply by ataffanel** [reply]
The fact that the red led on the radio if flashing a low is not ideal: in a good environment the green LED is very solid and the red one should blink only occasional.

Do you get any different blinking pattern on the radio when connecting a single Crazyflie with one radio?

One thing we have seen in our flight arena is that we have a very radio-reflective floor, this makes the radio communication very bad when Crazyflies are landed directly on the floor. We can make it a log better by lifting the Crazyflie a couple of centimeters to a decimeter.

Another thing I can think about is to space the channels: Kristoffer's example channels where spaced by 20MHz and it is not completely at random. Channels must be spaced by 2MHz minimum, but more is better to avoid radios to affects each others. Similarly it is good to keep at least 10cm between the radio dongle antennas to avoid having them pushing high-power radio signal into each-other.

**Reply by MeghanHailey** [reply]
Thanks for your reply and all the suggestions.
The blinking pattern seems about the same when connecting to a single Crazyflie with a single radio.
I have tried lifting them or even putting them on a table and it doesn't seem to make any difference to the blinking, so I don't think the floor is the issue in my case, or at least not the only issue.
I've been using very spaced out channels, they were something like 2, 60 and 110 so hopefully that shouldn't create interference. With my current setup I can only really separate 2 of the 3 radios I'm using by 10cm but will try getting a USB extender to test if that makes any difference and play around with it.

I think however I have found the issue and it's unrelated to all of this. It seems when I run the client or my scripts outside of the virtual machine the connection is a lot more stable - lot less red flashing, quicker cfclient connection times, less disconnecting and a lot shorter delay at the start of running the swarm script even with multiple drones. I don't know much about virtual machines, what could be causing this or what I could do to try fix it, but I think for now my best bet is not using the virtual machine unless you have any suggestions for things to try. However, when I do run a script outside of the virtual machine on the host system, for some reason I get this error despite using the same version of cflib: `AttributeError: 'Swarm' object has no attribute 'reset_estimators'`. I would appreciate any suggestion on how to fix this as otherwise the rest of the code works fine.
 I know this is now unrelated, so let me know if I should create a new post for this.

**Reply by ataffanel** [reply, marked as answer]
Ho, I did not realized you where running from the virtual machine (my bad for not reading the history carefully). Yes, running outside the virtual machine is definitely advised when trying to control a swarm: for some reason the way USB is handled by virtual machines is quite slow, which is alright to get started quickly with Crazyflie but causes problems for more complex experiments.

As for the error, you should double-check the version you are using, trying to upgrade it with `pip install cflib --upgrade`. The method [reset_estimators](https://github.com/bitcraze/crazyflie-lib-python/blob/0.1.23/cflib/crazyflie/swarm.py#L179) is present in the latest version of the cflib (0.1.23).

**Reply by MeghanHailey** [reply]
Thanks for confirming the virtual machine issues are known. I will avoid using it for swarm control from now. I have double checked I am running version 0.1.23 of cflib, do you have any further suggestions for what I could try to resolve the issue?

**Reply by ataffanel** [reply]
Could you send a minimal script that I could try here that exhibit the problem? This is very odd since the function should be there in this version, my only other ideas would be to uninstall and re-install the lib with pip.

**Reply by MeghanHailey** [reply]
I think I worked out the issue now. I tried running it in terminal and it worked, turned out VS Code was set to run a different instance of Python than the one I was installing the library into. Will need to look through all I have installed (it's a work laptop I "inherited" after someone else) but I think I've got the Anaconda installation selected in VS Code now and it's working well as that is the installation the pip command seems to be using. Thanks for all your help.

**Reply by ataffanel** [reply]
Great that this misterry is solved! Yes, this is the kind of problem I was expecting, happened to me multiple time and it is always confusing :-).
